### PR TITLE
DEX-789 Better error handling due Kafka issues

### PR DESCRIPTION
**Pull request by vsuharnikov**
Fixed a potential bug with doubled orders:
* Changed semantics to At Most Once;
* Introduced a circuit breaker behind the queue;
* Added failedPlacements to AddressActor (see comments);

Fixed other bugs: 
* Orders those is cancelling doesn't cancel again;
* Ignore cancelling orders during low balance cancels.

Other:
* Little readability improvements.

Tests:
* Added new property based tests for OrderBook;
* MultipleMatchersOrderCancelTestSuite now is correct;
* Improved KafkaIssuesTestSuite, now it is correct.

### CODE REVIEW

File: dex/src/main/scala/com/wavesplatform/dex/AddressActor.scala [hunk 1]
```diff
@@ -442,11 +454,12 @@ class AddressActor(owner: Address,
   }
 
   private def getOrdersToCancel(actualBalance: Map[Asset, Long]): Queue[InsufficientBalanceOrder] = {
+    val inProgress = pendingCancels
     // Now a user can have 100 active transaction maximum - easy to traverse.
     activeOrders.values.toVector
       .sortBy(_.order.timestamp)(Ordering[Long]) // Will cancel newest orders first
       .iterator
-      .filter(_.isLimit)
+      .filter(x => x.isLimit && !inProgress.contains(x.id))
```
Comment (DenisSerebryansky) [inline dex/src/main/scala/com/wavesplatform/dex/AddressActor.scala:462]: _Apply_ method of immutable set already returns boolean

File: dex/src/main/scala/com/wavesplatform/dex/Matcher.scala [hunk 2]
```diff
@@ -180,14 +180,24 @@ class Matcher(settings: MatcherSettings)(implicit val actorSystem: ActorSystem)
...
+  private val storeBreaker = new CircuitBreaker(
+    actorSystem.scheduler,
+    maxFailures = settings.eventsQueue.circuitBreaker.maxFailures,
+    callTimeout = settings.eventsQueue.circuitBreaker.callTimeout,
+    resetTimeout = settings.eventsQueue.circuitBreaker.resetTimeout
+  )
+
+  private def storeEvent(payload: QueueEvent): Future[Option[QueueEventWithMeta]] =
+    storeBreaker.withCircuitBreaker(matcherQueue.storeEvent(payload))
+
```
Comment (DenisSerebryansky) [inline dex/src/main/scala/com/wavesplatform/dex/Matcher.scala:192]: :+1:

File: dex/src/test/scala/com/wavesplatform/dex/model/OrderBookSpec.scala [hunk 3]
```diff
@@ -158,6 +157,35 @@ ${diff.mkString("\n")}
...
+
+OrderBook before:
+${format(ob)}
+
+OrderBook after:
+${format(updatedOb)}
+
+Events:
+${formatEvents(events)}
+"""
```
Comment (DenisSerebryansky) [inline dex/src/test/scala/com/wavesplatform/dex/model/OrderBookSpec.scala:181]:
```
Бе ды
   с Верст кой
```

File: dex/src/test/scala/com/wavesplatform/dex/model/OrderBookSpec.scala [hunk 4]
```diff
@@ -322,6 +354,12 @@ ${canceledOrders.mkString("\n")}
     }
   }
 
+  private def obInvariant(ob: OrderBook): Unit = {
+    val firstAskPrice = ob.asks.headOption.map(_._1).getOrElse(Long.MaxValue)
+    val firstBidPrice = ob.bids.headOption.map(_._1).getOrElse(Long.MinValue)
+    firstAskPrice should be > firstBidPrice
+  }
+
```
Comment (DenisSerebryansky) [inline dex/src/test/scala/com/wavesplatform/dex/model/OrderBookSpec.scala:362]: Bad name imho
What about `assertSpreadSign`?

File: dex/src/test/scala/com/wavesplatform/dex/settings/BaseSettingsSpecification.scala [hunk 5]
```diff
@@ -155,6 +155,12 @@ class BaseSettingsSpecification extends AnyFlatSpec {
          |          client.bar = 3
          |        }
          |      }
+         |
+         |      circuit-breaker {
+         |        max-failures = 999
+         |        call-timeout = 123s
+         |        reset-timeout = 1d
```
Comment (DenisSerebryansky) [inline dex/src/test/scala/com/wavesplatform/dex/settings/BaseSettingsSpecification.scala:162]: Do you have some tests for circuit-breaker? 
I can suggest unit test in `MatcherApiRouteSpecification`

File: dex-it/src/test/scala/com/wavesplatform/it/sync/KafkaIssuesTestSuite.scala [hunk 6]
```diff
@@ -106,4 +101,40 @@ class KafkaIssuesTestSuite extends WsSuiteBase with HasWebSockets with HasKafka
...
+
+  private def disconnectKafkaFromNetwork(): Unit = {
+    writeGlobalLog("--- Disconnecting Kafka from the network ---")
+    kafka.dockerClient
+      .disconnectFromNetworkCmd()
+      .withContainerId(kafka.containerId)
+      .withNetworkId(networkId)
+      .exec()
+
+    waitForNetworkSettings(_.getNetworks.containsKey(networkId))
```
[the remark this answers is not in the document]
Comment (vsuharnikov) [reply dex-it/src/test/scala/com/wavesplatform/it/sync/KafkaIssuesTestSuite.scala:113]: Yep :) Predicates in `disconnectKafkaFromNetwork` and `connectKafkaToNetwork` should be swapped

File: dex-it/src/test/scala/com/wavesplatform/it/sync/networking/MultipleMatchersOrderCancelTestSuite.scala [hunk 7]
```diff
@@ -47,15 +49,25 @@ class MultipleMatchersOrderCancelTestSuite extends MatcherSuiteBase {
 
     dex1.api.saveSnapshots
     dex1.restartWithNewSuiteConfig(ConfigFactory.parseString(s"waves.dex.events-queue.type = local").withFallback(dexInitialSuiteConfig))
+    // HACK: Because we switched the queue, we need to place 5 orders to move offset of queue
+    (1 to 5).foreach { i =>
+      val order = mkOrderDP(acc3, ethWavesPair, OrderType.SELL, 1.eth, 1)
+      dex1.api.place(order)
+    }
```
[the remark this answers is not in the document]
Comment (vsuharnikov) [reply dex-it/src/test/scala/com/wavesplatform/it/sync/networking/MultipleMatchersOrderCancelTestSuite.scala:56]: You're right :)

File: dex/src/main/scala/com/wavesplatform/dex/AddressActor.scala (hunk 1 above)
Comment (vsuharnikov) [reply dex/src/main/scala/com/wavesplatform/dex/AddressActor.scala:462]: Don't get what did you mean :) If you are speaking about `inProgress.contains` vs `inProgress`, the first version is more readable.

File: dex/src/test/scala/com/wavesplatform/dex/model/OrderBookSpec.scala (hunk 4 above)
Comment (vsuharnikov) [reply dex/src/test/scala/com/wavesplatform/dex/model/OrderBookSpec.scala:362]: An invariant is a property that should be always held. I see here only two properties:
1. Asks and bids have right order;
2. Asks and bids don't intersect each other.

The first is guaranteed by TreeMap. so there is only one invariant. Changed to `obAskBidIntersectionInvariant`

File: dex/src/test/scala/com/wavesplatform/dex/settings/BaseSettingsSpecification.scala (hunk 5 above)
Comment (vsuharnikov) [reply dex/src/test/scala/com/wavesplatform/dex/settings/BaseSettingsSpecification.scala:162]: The circuit breaker we are using is implemented and well [tested](https://github.com/akka/akka/blob/master/akka-actor-tests/src/test/scala/akka/pattern/CircuitBreakerSpec.scala) in Akka.

Will add an integration test that checks we can continue work when Kafka issues are solved.

File: dex-it/src/test/scala/com/wavesplatform/it/sync/KafkaIssuesTestSuite.scala [hunk 8]
```diff
@@ -37,7 +52,31 @@ class KafkaIssuesTestSuite extends WsSuiteBase with HasWebSockets with HasKafka
...
+    Thread.sleep(resetTimeout.toMillis)
+
+    connectKafkaToNetwork()
+    Thread.sleep(waitAfterNetworkChanges.toMillis)
+
+    withClue("Messages weren't saved to the queue") {
+      dex1.api.currentOffset shouldBe offsetBefore
+    }
+
+    // dex1.api.tryPlace(mkOrderDP(alice, wavesUsdPair, SELL, 1.waves, 3.0)) shouldBe 'right
```
Comment (DenisSerebryansky) [inline dex-it/src/test/scala/com/wavesplatform/it/sync/KafkaIssuesTestSuite.scala:73]: :thinking:

**Pull request merged**